src/ir_passes/assume.rs

```diff
@@ -0,0 +1,174 @@
+                .implies(self.add_orderconstraint(cons, binding), self),
+        }
+    }
+}
+
+impl Assume {
+    fn prop(
+        p: ir::PropIdx,
+        info: ir::InfoIdx,
+        comp: &mut ir::Component,
+    ) -> Action {
+        log::debug!("Checking assumptions for {p}");
+        let p = comp.get(p);
+        match p {
+            ir::Prop::Cmp(ir::CmpOp {
+                op: ir::Cmp::Eq,
+                lhs,
+                rhs,
+            }) => {
+                let exprs = comp.exprs();
+                let funcmatch = match (exprs.get(*lhs), exprs.get(*rhs)) {
+                    (ir::Expr::Fn { op, args }, _) => {
+                        debug_assert!(args.len() == 1, "Currently Unimplemented: {} requires {} arguments, automatic assumptions only implemented for single argument functions.", op, args.len());
+                        Some((
+                            op,
+                            Binding::new(vec![
+                                (FnAssume::left(), *rhs),
+                                (FnAssume::right(), args[0]),
+                            ]),
+                        ))
+                    }
+                    (_, ir::Expr::Fn { op, args }) => {
+                        debug_assert!(args.len() == 1, "Currently Unimplemented: {} requires {} arguments, automatic assumptions only implemented for single argument functions.", op, args.len());
+                        Some((
+                            op,
+                            Binding::new(vec![
+                                (FnAssume::left(), *lhs),
+                                (FnAssume::right(), args[0]),
+                            ]),
+                        ))
+                    }
+                    _ => None,
+                };
+
+                match funcmatch {
+                    None => Action::Continue,
+                    Some((op, binding)) => {
+                        log::debug!("Matched function");
+                        Action::AddBefore(
+                            FnAssume::from(op.clone())
+                                .assumptions
+                                .into_iter()
+                                .filter_map(|assumption| {
+                                    let imp = comp
+                                        .add_implication(assumption, &binding);
+                                    log::debug!("Assuming {}", imp);
+                                    comp.assume(imp, info)
+                                })
+                                .collect(),
+                        )
+                    }
+                }
+            }
+            ir::Prop::And(lhs, rhs) => {
+                let (lhs, rhs) = (*lhs, *rhs);
+                let laction = Assume::prop(lhs, info, comp);
+                match laction {
+                    Action::AddBefore(cmds) => {
+                        match Assume::prop(rhs, info, comp) {
```

Comment:
I'm not sure what's going on here? Why does this need to call a function that returns a Visitor action?

Reply:
Yeah I'm probably gonna restructure this looking back, the adds should be moved elsewhere.